### ISSUE TITLE
Topic5: Coroutine & Asynchronous programming (协程&异步编程)

### ISSUE THREAD

**Issue by ketoo** [ketoo/NoahGameFrame#74]
Background:

Normally, the program executes a sequence of computations by line. Sometimes, the codes will spend a lot of time in reading files or waiting for something. Is there have any way to optimize this situation?

Please leave messages here for discussing.

**Comment by MacroGu** [comment]
As showed before, does the load file or wait for some task finished required by the program to run, if it is, I think there is no way to optimize it, but I suggest that, we could load all file and some necessary task when the program initialize, and do some normal operations of storing data to container or parsing. use big memory and a lot of time during start up initialize to get a faster and easy program running later.

if not, we could use observer Pattern to solve this program, load file module hold an instance of the main module, after load file finished or task is finished, use that instance to notice main module do other thing.

上面说的加载文件和等待某个任务完成，是不是程序运行下去的必要条件， 如果是，我觉得没法优化， 但是建议程序在开始运行的时候，初始化加载全部需要的文件， 并且做一些固定的存储容器和解析的操作。用大的内存来和长的初始化时间换取程序后面运行的方便和快速。

如果不是，是不是可以用 observer 模式来解决这个问题， 加载文件的模块保留一个 主模块的对象， 在完成加载或者任务完成之后，再通知主模块进行其他的任务。

**Comment by ketoo** [comment]
@MacroGu Yes, basically, people loaded all resources that the application required when launching the application.

Observer Pattern you mentioned is a good way to solve the problem that some tasks blocking the main thread. The disadvantage of this way is it needs another callback function to receive the notice when the task finished, which is unfriendly in terms of synchronizing code.

Actually, some coder hopes all the business code run in one thread and with a high performance, besides, the function can transfer its context and can restore its context when the program needs.

The answer is Coroutine.

The project([](https://github.com/ketoo/NFCoroutine)) now I am writing is a demo about Coroutine. Another project ([](https://github.com/cloudwu/coroutine)) writen by Cloud.Wu is the same kind of project.
